tests/test_patch_rebaser.py

```diff
@@ -41,3 +45,254 @@ def test_parse_distro_info_path():
+    assert mock_repo.git.push.called is True
+
+    expected = [(("my_remote", "private-rebaser-2019-previous"),),
+                (("-f", "my_remote", "my_branch"),)]
+    assert mock_repo.git.push.call_args_list == expected
+
+
+def test_perform_rebase(mock_repo):
+    """
+    GIVEN Rebaser initialized correctly including branch and commit
+    WHEN perform_rebase is called
+    THEN branch.rebase_to_hash is called
+    WITH the same branch and commit
+    """
+    rebaser = Rebaser(mock_repo, "my_branch", "my_commit", "my_remote",
+                      "my_tstamp", dev_mode=True)
+    rebaser.perform_rebase()
+
+    mock_repo.branch.rebase_to_hash.assert_called()
+    mock_repo.branch.rebase_to_hash.assert_called_with(
+        "my_branch", "my_commit"
+    )
+
+
+def test_perform_rebase_aborts_on_failure(mock_repo):
+    """
+    GIVEN Rebaser initialized correctly
+    WHEN rebase_to_hash fails with RebaseException
+    THEN perform_rebase also raises RebaseException
+    AND abort_rebase is called
+    """
+    mock_repo.branch.rebase_to_hash.side_effect = exceptions.RebaseException
+
+    rebaser = Rebaser(mock_repo, "my_branch", "my_commit", "my_remote",
+                      "my_tstamp", dev_mode=True)
+
+    with pytest.raises(exceptions.RebaseException):
+        rebaser.perform_rebase()
+
+    mock_repo.branch.abort_rebase.assert_called()
+
+
+def test_rebase_and_update_remote(mock_repo, monkeypatch):
+    """
+    GIVEN Rebaser initialized correctly
+    WHEN rebase_and_update_remote is called
+    THEN create_tag is called once
+    AND remote.fetch is called twice
+    AND git.push is called
+    """
+    remote, branch, remote_sha = "my_remote", "my_branch", "abc"
+
+    monkeypatch.setattr(time, 'sleep', lambda s: None)
+
+    # TODO: Mock more cleanly once git_wrapper updated with comparison function
+    mock_branch = Mock(**{'commit.hexsha': remote_sha})
+    mock_remote = Mock(refs={branch: mock_branch})
+    mock_tag = Mock(**{'commit.hexsha': remote_sha})
+    mock_repo.repo = Mock(remotes={remote: mock_remote},
+                          **{'create_tag.return_value': mock_tag})
+
+    rebaser = Rebaser(mock_repo, "my_branch", "my_commit", "my_remote",
+                      "2019", dev_mode=True)
+    rebaser.rebase_and_update_remote()
+
+    assert mock_repo.repo.create_tag.call_count == 1
+    assert mock_repo.remote.fetch.call_count == 2
+
+    expected = [(("-n", "my_remote", "private-rebaser-2019-previous"),),
+                (("-nf", "my_remote", "my_branch"),)]
+    assert mock_repo.git.push.call_args_list == expected
+
+
+def test_rebase_and_update_remote_success_after_retry(mock_repo, monkeypatch):
+    """
+    GIVEN Rebaser initialized correctly
+    WHEN rebase_and_update_remote is called
+    AND the remote changes once during the rebase
+    THEN create_tag is called twice
+    AND remote.fetch is called four times
+    AND git.tag is called once for deleting the first tag
+    AND git.push is called
```

Comment:
Ok, I am going to approve this patch, as the comments you added are very helpful.  However, I would like to point out a minor detail in how to approach these descriptions, as I have seen several of these follow the above pattern, and I think it will become onerous to keep up.  These descriptions are really meant to be a bit higher level, describing a scenario, and the expected outcome.  Generally you don't need to include things like how many times a certain command is called, as that is handled by the test itself.  Instead, you just want to describe the expected outcome.  For example, with the above comment, you could say something along the lines of:
```
Given a valid Rebaser that has called rebase_and_update_remote
When the remote changes once during the rebase
Then Rebaser should catch that there has been a change,
And tag and push the rebased branch.
```
(Assuming I have understood the correct behavior.  This is just an example, you can word it however you want - the point is to describe the _behavior_ rather than the _implementation_.  However, I didn't ask you to add these to increase the workload, so I am happy to have them as is, this is just an explanation to hopefully clarify my intent.

Reply:
That is helpful, thank you. I seem to recall when git_wrapper was being created that the idea was that eventually, we could use a tool to automatically check/tests these comments directly. If we're free to be verbose in our descriptions though, well, that definitely would be useful in more complex cases like the ones here so I'd be happy to do it that way instead!